src/server.rs

```diff
@@ -0,0 +1,512 @@
+                        debug!("Cid {cid} not in blockstore for {peer}");
+                    }
+                    BlockstoreResult::Get(peer, cid, Ok(Some(data))) => {
+                        trace!("Cid {cid} for {peer} present in blockstore");
+                        self.outgoing_queue.push((cid, data));
+                    }
+                    BlockstoreResult::Get(_peer, cid, Err(error)) => {
+                        warn!("Fetching {cid} from blockstore failed: {error}");
+                    }
+                    BlockstoreResult::Cancelled => (),
+                }
+                continue;
+            }
+
+            if self.update_handlers() {
+                continue;
+            }
+
+            return Poll::Pending;
+        }
+    }
+}
+
+pub(crate) struct ServerConnectionHandler<const S: usize> {
+    protocol: StreamProtocol,
+    sink: SinkState,
+    pending_outgoing_messages: Option<Vec<ProtoBlock>>,
+}
+
+impl<const S: usize> fmt::Debug for ServerConnectionHandler<S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("ServerConnectionHandler")
+    }
+}
+
+#[derive(Default)]
+enum SinkState {
+    #[default]
+    None,
+    Requested,
+    Ready(Sink),
+}
+
+impl<const S: usize> ServerConnectionHandler<S> {
+    pub(crate) fn set_stream(&mut self, stream: libp2p_swarm::Stream) {
+        // Convert `AsyncWrite` stream to `Sink`
+        self.sink = SinkState::Ready(FramedWrite::new(stream, Codec));
+    }
+
+    pub(crate) fn queue_messages(&mut self, messages: Vec<(Vec<u8>, Vec<u8>)>) {
+        let block_list = messages
+            .into_iter()
+            .map(|(prefix, data)| ProtoBlock { prefix, data })
+            .collect::<Vec<_>>();
+
+        self.pending_outgoing_messages
+            .get_or_insert(Vec::with_capacity(block_list.len()))
+            .extend(block_list);
+    }
+
+    fn open_new_substream(
+        &mut self,
+    ) -> Poll<
+        ConnectionHandlerEvent<ReadyUpgrade<StreamProtocol>, StreamRequester, ToBehaviourEvent<S>>,
+    > {
+        self.sink = SinkState::Requested;
+
+        Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+            protocol: SubstreamProtocol::new(
+                ReadyUpgrade::new(self.protocol.clone()),
+                StreamRequester::Server,
+            ),
+        })
+    }
+
+    fn poll_outgoing(
+        &mut self,
+        cx: &mut Context,
+    ) -> Poll<
+        ConnectionHandlerEvent<ReadyUpgrade<StreamProtocol>, StreamRequester, ToBehaviourEvent<S>>,
+    > {
+        loop {
+            match (&mut self.pending_outgoing_messages, &mut self.sink) {
+                (_, SinkState::Requested) => return Poll::Pending,
+                (None, SinkState::None) => return Poll::Pending,
+                (None, SinkState::Ready(sink)) => {
+                    if ready!(sink.poll_flush_unpin(cx)).is_err() {
+                        self.close_sink("poll_flush_unpin");
+                    }
```

Comment:
after `flush` is finished we should close the sink if there is nothing else to send

Reply:
Added dropping sink after flush 👍